src/server.rs

```diff
@@ -0,0 +1,352 @@
+    pub fn bind_secure<S: Into<String>>(mut self,
+                                        host: S,
+                                        port: u32,
+                                        c: ServerCredentials)
+                                        -> ServerBuilder {
+        self.addrs.push((host.into(), port, Some(c)));
+        self
+    }
+
+    /// Add additional configuration for each incoming channel.
+    pub fn channel_args(mut self, args: ChannelArgs) -> ServerBuilder {
+        self.args = Some(args);
+        self
+    }
+
+    /// Set how many requests a completion queue can handled.
+    pub fn requests_slot_per_cq(mut self, slots: usize) -> ServerBuilder {
+        self.slots_per_cq = slots;
+        self
+    }
+
+    /// Register a service.
+    pub fn register_service(mut self, service: Service) -> ServerBuilder {
+        self.handlers.extend(service.handlers);
+        self
+    }
+
+    pub fn build(mut self) -> Server {
+        let args = self.args.map_or_else(ptr::null, |args| args.as_ptr());
+        unsafe {
+            let server = grpc_sys::grpc_server_create(args, ptr::null_mut());
+            let bind_addrs: Vec<_> = self.addrs
+                .drain(..)
+                .map(|(host, port, certs)| {
+                    let addr = format!("{}:{}\0", host, port);
+                    let addr_ptr = addr.as_ptr();
+                    let bind_port = match certs {
+                        None => {
+                            grpc_sys::grpc_server_add_insecure_http2_port(server, addr_ptr as _)
+                        }
+                        Some(mut cert) => {
+                            grpc_sys::grpc_server_add_secure_http2_port(server,
+                                                                        addr_ptr as _,
+                                                                        cert.as_mut_ptr())
+                        }
+                    };
+
+                    (host, bind_port as u32)
+                })
+                .collect();
+
+            for cq in self.env.completion_queues() {
+                grpc_sys::grpc_server_register_completion_queue(server,
+                                                                cq.as_ptr(),
+                                                                ptr::null_mut());
+            }
+
+            Server {
+                inner: Arc::new(Inner {
+                                    env: self.env,
+                                    server: server,
+                                    shutdown: AtomicBool::new(false),
+                                    bind_addrs: bind_addrs,
+                                    slots_per_cq: self.slots_per_cq,
+                                    handlers: self.handlers,
+                                }),
+            }
+        }
+    }
+}
+
+pub struct Inner {
+    env: Arc<Environment>,
+    server: *mut GrpcServer,
+    bind_addrs: Vec<(String, u32)>,
```

Comment:
Agree with `u16`, but not `SocketAddr`. I think `String` is for `:authority`, e.g. `:authority = pubsub.googleapis.com`.